presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java

```diff
@@ -216,47 +218,65 @@ public void containerStopping(DockerContainer container, InspectContainerRespons
     static EnvironmentListener statsPrintingListener()
     {
         ScheduledExecutorService executorService = Executors.newScheduledThreadPool(2, daemonThreadsNamed("container-stats-%d"));
-        List<ScheduledFuture<?>> futures = new ArrayList<>();
+        Map<String, StatisticsFetcher> fetchers = new TreeMap<>(); // keep map sorted
 
         return new EnvironmentListener()
         {
             @Override
-            public void containerStarting(DockerContainer container, InspectContainerResponse response)
+            public void containerStarted(DockerContainer container, InspectContainerResponse response)
             {
-                // Print stats every 30 seconds
-                futures.add(executorService.scheduleWithFixedDelay(() ->
-                {
-                    StatisticsFetcher.Stats stats = container.getStats();
-                    if (stats.areCalculated()) {
-                        log.info("%s - %s", container.getLogicalName(), container.getStats());
-                    }
-                }, 5 * 1000L, 30 * 1000L, MILLISECONDS));
+                // Start listening on statistics stream
+                fetcher(container).start();
             }
 
             @Override
-            public void containerStopping(DockerContainer container, InspectContainerResponse response)
+            public void environmentStopping(Environment environment)
             {
-                log.info("Container %s final statistics - %s", container, container.getStats());
-            }
+                log.info("Container final stats: ");
+                printContainerStats();
```

Comment:
nit: closing and then printing seems more natural.

Reply:
makes sense